sklearn/tree/_splitter.pyx

```diff
@@ -42,7 +42,7 @@ cdef inline void _init_split(SplitRecord* self, SIZE_t start_pos) nogil:
     self.threshold = 0.
     self.improvement = -INFINITY
 
-cdef class Splitter:
+cdef class BaseSplitter:
```

Comment:
We should add docstring similar to what is in Criterion. So something along these lines:

```
This is an abstract interface for splitters. 

For example, a tree model could be either supervisedly, or unsupervisedly computing splits on samples of
covariates, or labels, or both. Although scikit-learn currently only contains
supervised tree methods, this class enables 3rd party packages to leverage
scikit-learn's Cython code for splitting. 

A splitter is usually used in conjunction with a criterion class, which explicitly handles
computing the criteria, which we split on.

The downstream classes _must_ implement methods to compute the split in a node.
```